### PR TITLE
[train] Default to chunked logprobs for Megatron

### PR DESCRIPTION
**Pull request by SumanthRH**
# What does this PR do?

Defaults to chunked logprobs for Megatron with chunk_size of 1024

Currently, we don't do any chunking for logprobs calculation in the Megatron backend. For models like Qwen 3.5 with large vocab sizes (248,320), this can lead to OOMs. 

This PR chooses a chunk size of 1024 as the best balance for memory saved vs increase in latency. For large sequence lengths and vocab sizes, this can save a lot of memory - ex: for 65K sequence lenght and 128K vocab size, we save 42GB of memory while the time taken  increases from 132.2 ms to 600 ms (acceptable IMO). 

## Current memory usage

Currently, the full logprobs tensor of size `(bsz, seq_len, vocab_size)` gets materialized while we calculate the logprobs for the given input_ids. 

The only way users can save memory in logprobs calculation is with tensor parallelism right now, where each TP rank would work with Tp sharded logits `(bsz, seq_len, vocab_size/ TP)` . Chunking along sequence is also needed for large sequence lengths.

## Benchmark Settings

**Device:** NVIDIA B200
**Vocab sizes:** 32,000, 64,000 and 128,000
**Measurement:** forward + backward pass
**Chunk sizes tested:** None (no chunking), 32, 1024, 4096, 8192, 16384. I added 32 chunk size to the mix just to see what a small chunk size will yield.

### Vocab size 32K

| seq_len | chunk_size | time (ms) | peak mem (MB) | vs no-chunk | mem saved (MB) |
|--------:|-----------:|----------:|--------------:|------------:|---------------:|
|  32,768 |       None |      17.0 |        15,601 |       1.00x |              0 |
|  32,768 |         32 |     311.3 |        10,016 |       0.05x |          5,585 |
|  32,768 |      1,024 |      41.5 |        10,501 |       0.41x |          5,100 |
|  32,768 |      4,096 |      38.3 |        12,001 |       0.44x |          3,600 |
|  32,768 |      8,192 |      37.7 |        14,001 |       0.45x |          1,600 |
|  32,768 |     16,384 |      37.5 |        20,001 |       0.45x |         -4,400 |
|  65,536 |       None |      33.5 |        31,202 |       1.00x |              0 |
|  65,536 |         32 |     627.8 |        20,017 |       0.05x |         11,185 |
|  65,536 |      1,024 |      82.7 |        20,501 |       0.41x |         10,700 |
|  65,536 |      4,096 |      76.5 |        22,001 |       0.44x |          9,200 |
|  65,536 |      8,192 |      75.2 |        24,001 |       0.45x |          7,200 |
|  65,536 |     16,384 |      74.7 |        28,001 |       0.45x |          3,200 |
| 131,072 |       None |      66.6 |        62,404 |       1.00x |              0 |
| 131,072 |         32 |   1,269.6 |        40,018 |       0.05x |         22,385 |
| 131,072 |      1,024 |     165.9 |        40,503 |       0.40x |         21,901 |
| 131,072 |      4,096 |     152.9 |        42,003 |       0.44x |         20,401 |
| 131,072 |      8,192 |     150.2 |        44,003 |       0.44x |         18,401 |
| 131,072 |     16,384 |     149.0 |        48,003 |       0.45x |         14,401 |

### Vocab Size 64K

| seq_len | chunk_size | time (ms) | peak mem (MB) | vs no-chunk | mem saved (MB) |
|--------:|-----------:|----------:|--------------:|------------:|---------------:|
|  32,768 |       None |      33.4 |        31,201 |       1.00x |              0 |
|  32,768 |         32 |     316.1 |        20,032 |       0.11x |         11,169 |
|  32,768 |      1,024 |      78.6 |        21,001 |       0.42x |         10,200 |
|  32,768 |      4,096 |      75.0 |        24,001 |       0.45x |          7,200 |
|  32,768 |      8,192 |      74.5 |        28,001 |       0.45x |          3,200 |
|  32,768 |     16,384 |      74.5 |        40,001 |       0.45x |         -8,800 |
|  65,536 |       None |      66.4 |        62,402 |       1.00x |              0 |
|  65,536 |         32 |     637.5 |        40,033 |       0.10x |         22,369 |
|  65,536 |      1,024 |     157.4 |        41,001 |       0.42x |         21,400 |
|  65,536 |      4,096 |     149.8 |        44,001 |       0.44x |         18,400 |
|  65,536 |      8,192 |     148.6 |        48,001 |       0.45x |         14,400 |
|  65,536 |     16,384 |     148.5 |        56,001 |       0.45x |          6,400 |
| 131,072 |       None |     133.7 |       124,804 |       1.00x |              0 |
| 131,072 |         32 |   1,274.2 |        80,034 |       0.10x |         44,770 |
| 131,072 |      1,024 |     314.7 |        81,003 |       0.42x |         43,801 |
| 131,072 |      4,096 |     299.4 |        84,003 |       0.45x |         40,801 |
| 131,072 |      8,192 |     297.0 |        88,003 |       0.45x |         36,801 |
| 131,072 |     16,384 |     296.7 |        96,003 |       0.45x |         28,801 |

### Vocab Size 128K

| seq_len | chunk_size | time (ms) | peak mem (MB) | vs no-chunk | mem saved (MB) |
|--------:|-----------:|----------:|--------------:|------------:|---------------:|
|  32,768 |       None |      66.4 |        62,401 |       1.00x |              0 |
|  32,768 |         32 |     331.5 |        40,063 |       0.20x |         22,338 |
|  32,768 |      1,024 |     152.8 |        42,001 |       0.43x |         20,400 |
|  32,768 |      4,096 |     148.5 |        48,001 |       0.45x |         14,400 |
|  32,768 |      8,192 |     148.3 |        56,001 |       0.45x |          6,400 |
|  32,768 |     16,384 |     148.3 |        80,001 |       0.45x |        -17,600 |
|  65,536 |       None |     132.2 |       124,802 |       1.00x |              0 |
|  65,536 |         32 |     699.5 |        80,064 |       0.19x |         44,738 |
|  65,536 |      1,024 |     305.5 |        82,001 |       0.43x |         42,800 |
|  65,536 |      4,096 |     296.9 |        88,001 |       0.45x |         36,800 |
|  65,536 |      8,192 |     296.5 |        96,001 |       0.45x |         28,800 |
|  65,536 |     16,384 |     296.3 |       112,001 |       0.45x |         12,800 |
| 131,072 |       None |       OOM |             — |           — |              — |
| 131,072 |         32 |       OOM |             — |           — |              — |
| 131,072 |      1,024 |       OOM |             — |           — |              — |
| 131,072 |      4,096 |       OOM |             — |           — |              — |
| 131,072 |      8,192 |       OOM |             — |           — |              — |
| 131,072 |     16,384 |       OOM |             — |           — |              — |

### CODE REVIEW

File: skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py [hunk 1]
```diff
@@ -32,6 +32,11 @@
 from skyrl.backends.skyrl_train.utils.torch_utils import masked_mean
 from skyrl.train.config import TrainerConfig
 
+# NOTE (sumanthrh): We use a chunk size of 1024 for calculating logprobs
+# from logits to avoid OOMs for large sequence lengths.
+# For more details, see https://github.com/NovaSky-AI/SkyRL/pull/1610
+CHUNK_SIZE_LOGPROBS = 1024
```
[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py]: Added `logprobs_chunk_size` to trainer config
